ls1intum/Artemis
# [Exam] Test Run

### Checklist

### Motivation and Context
When instructors created exams, they have no way of testing their exam. Their only workaround was to create a new exam and register themselves as a student. We want to give the opportunity to instructors to create test runs to experience the exam as the students would.

### Description
**Distinguishing test runs from student exams**
We added a column `test_run` to flag student exams as test runs. Only Instructors are allowed to create test runs, for which they are registered as the `student`. 

**Distinguishing test run participation, submissions and results**
We also need a way to mark the student participation and its submissions which belong to test runs. This is important because we do not want these submissions to count against the statistics of the exam. Additionally, tutors should not be able to see or assess these submissions. Therefore, the chosen solution for this uses the student field of the participation and the assessor field of the result. 
When a test run is created for specific exercises, as part of the setup process, an initial submission is created (similar to the normal exam participations). Additionally, a result is created where the assessor is set using the same user. Therefore we create the condition: `student_id = assessor_id`. 
If a `participation.exercise.hasExerciseGroup()` (is an exam exercise) and there is at least one submission which fulfils the condition: `participation.id = submission.result.assessor.id`, then we can be sure that this is a test run submission and we can ignore it.
The condition is set to at least one, as programming exercises can contain multiple submissions. However, the initial one will ensure that it is marked as a test run exercise. 

**Quirks**
Instructors can create multiple test runs for exams, assess their submissions through the test run dashboard, submit complaints about their assessment and respond to these complaints in order to simulate a full cycle. Test runs can be created before, during or after an exam. They have no influence on the exam runtime (`latestIndividualEndDate`), exam statistics, tutor exam dashboard statistics or results. If an instructor creates multiple test runs which reference the same exercise, the same participation is used for all of them. If a test run is deleted, and its exercise participation is referenced by another test run, the participation is not deleted. Otherwise the test run participations are deleted, including all build plans and repositories. 

**New API Endpoints:** 
Create a test run: 					    `POST /courses/{courseId}/exams/{examId}/create-test-run`
Delete a test run: 					    `POST /courses/{courseId}/exams/{examId}/delete-test-run/{testRunId}`
Get all Test Runs for Exam: 		            `GET /courses/{courseId}/exams/{examId}/test-runs`
Get interesting exercises for test run dashboard: `GET /courses/:courseId/exams/:examId:for-exam-tutor-test-run-dashboard`
Get complaints for test run dashboard: 	    `GET /exercises/{exerciseId}/complaints-for-test-run-dashboard`
Get test run for Conduction: 			    `GET /courses/{courseId}/exams/{examId}/test-run/{testRunId}/conduction`
Get all testruns for exam: 			    `GET /courses/{courseId}/exams/{examId}/test-runs`
Get all test run submissions: 			    `GET /exercises/{exerciseId}/test-run-submissions`

### Steps for Testing
**The most important thing to test is that nothing is broken for real exams**
Create an exam, register students, conduct and assess. Try out all features. Everything should work as expected.

**Testing test runs**
**Part one: Test Run**
1. **Log in as an administrator of a course** & create an exam add exercises and register students
_(do not create a test run for an exam where you are already registered as a student and have conducted an exam, this leads to unspecified behaviour)_
2. Create a test runs for exercises. 
3. Run the test run and submit. 
_Check the status bar for programming exercises. It should not default to a checkmark. It should only show submitted, once you have really submitted at least once!_
4. Assess the test run from the test run dashboard
5. Complain about the assessment from the summary of the test run 
6. Respond to your own complaint

**Part two: Add Real Student Exams**

7. Register students for the exam and run the conduction
8. assess the student submissions from the tutor course dashboard. The statistics should not display any information about the test run submissions
9. release the results and allow student reviews
10. try out the complain feature
11. Respond to the complains. 
12. Check out the scores view of the exam. It should only render results from real student exams. 

--------------------
### Test Coverage
#### Server:
ExamResource: 99%
StudentExamResource: 93%
TextSubmissionResource: 93%
ModelingSubmissionResource: 98%
ComplaintResource: 78%
ResultResource 87%
StudentExamService: 75%
ExamService: 84%
SubmissionService: 92%
#### Client: 
create-test-run-modal.component.ts 64%
test-run-management.component.ts 77%
assessment-header.component.ts 100%
assessment-layout.component.ts 98%

--------------------

### Screenshots 
**Create test run Modal**
Once you select an exercise, it is highlighted in Artemis blue to indicate that it will be part of the test run. Only once an exercise has been selected for every exercise group can the user create the test run.
![image](https://user-images.githubusercontent.com/15110960/93890001-2a460480-fcea-11ea-8742-faa98c3d6dfd.png)
**Test Run management screen**
Here you can manage the test runs and assess the test run submissions. These will not be visible from the other dashboards
![image](https://user-images.githubusercontent.com/15110960/93890312-827d0680-fcea-11ea-87c2-b8e5c0f7e133.png)
